### ISSUE TITLE
Consider using callr to render in a separate process

### ISSUE THREAD

**Issue by jennybc** [rstudio/rmarkdown#1204]
Inspired by discussions such as this one from SO [about leakage between current R session and Rmd code](https://stackoverflow.com/questions/47284424/rstudio-1-1-383-and-rmarkdown-1-7-environment-enheritance).

Would it make sense to offer a way to tell `render()` to work in a fresh R process? I do that in reprex (actually it's hard-wired that way), but don't have a good handle on all the implications of doing this for `render()` itself.

**Comment by yihui** [comment]
For reprex, it makes perfect sense -- users need to make sure their code works in a clean new R session. For rmarkdown, I don't know. I was surprised that the OP was surprised, because that has been the behavior of `rmarkdown::render()` since the very first version of rmarkdown (how could it be possible that he just discovered it yesterday, given that apparently he had used rmarkdown before?).

Anyway, I think this only matters to those who use rmarkdown programatically, in which case `callr::r(function() rmarkdown::render("test.Rmd"))` sounds easy enough to me, so I tend to leave it the users instead of introducing another dependency in rmarkdown and complexity to the `render()` function. Thanks for the suggestion!

**Comment by jennybc** [comment]
I think I also found this surprising at some point. And it's true that `rmarkdown::render()` is consistent with `source()` here re: two-way leakage. I imagine callr wasn't an option at the time the `envir` argument came into being, but I wonder if "run this in a fresh process" is essentially the goal when user requests to run in a new environment.

**Comment by yihui** [comment]
If we were to provide such an option in rmarkdown, the API would look like this:

```r
rmarkdown::render("test.Rmd", new_session = TRUE)
```

Without the option, the code looks like this:

```r
callr::r(function() rmarkdown::render("test.Rmd"))
```

For those who work with Rmd programmatically, I don't feel there is a huge advantage (OMG! The two lines have exactly the same number of characters!). From my perspective as the maintainer, I'd avoid introducing a new dependency (well, it is not only about implementation; I'll have to document it, maintain it, explain it to those who don't read the documentation, and possibly be buried in the debate on whether `new_session = TRUE` or `FALSE` is a better default). At the same time, as you mentioned, what does a new environment mean? It could be `new.env()`, or a new process. Either way, users can easily achieve what they want.

In short: it is something easy for users to do by themselves if they really care, but not so easy for this lazy maintainer. So I tend to [say no by default](https://37signals.com/rework).

![No, I'll just move on. Sorry.](https://slides.yihui.name/gif/latex-ignore.gif)

**Comment by gaborcsardi** [comment]
For the record, the naive way to invoke `render` via `callr` has a problem with S3 lookup. Specifically, S3 methods created in the vignetted are not found from package code. 

This is actually _not_ specific to `callr`, it is an interference between S3 lookup and calling  `render()` from within a function, or setting `envir`. 

With `callr` you can at least do this:

```r
path <- callr::r(
  function(...) rmarkdown::render(...),
  args = list(input, output_format, output_file, envir = globalenv())
)
```

Global env is fine here, because it is the global env of another process. I am not sure how to fix it without `callr`, though.

Here is an actual case when it came up:
https://github.com/hadley/pkgdown/issues/294

**Comment by yihui** [comment]
@gaborcsardi Thanks for the tip! I think this makes me even more reluctant to support Jenny's feature request. Too much magic for me to understand...
